File: 2021 카카오/메뉴 리뉴얼.cpp
#include<string>
#include<map>
#include<algorithm>
#include<vector>
using namespace std;

int arr[27];
vector<string> dp[27][21];
map<string, int> mp;

void find(string s1, int idx, string s2) {
    if(s2.size() > 1) {
        mp[s2]++;
        arr[s2.size()] = max(arr[s2.size()], mp[s2]);
        dp[s2.size()][mp[s2]].push_back(s2);
    }
    
    for(int i = idx + 1; i < s1.size(); i++) {
        s2.push_back(s1[i]);
        find(s1, i, s2);
        s2.pop_back();
    }
}

vector<string> solution(vector<string> orders, vector<int> course) {
    vector<string> answer;
    
    for(string &s : orders) {
        sort(s.begin(), s.end());
        find(s, -1, "");
    }
    
    for(int i : course) {
        if(arr[i] > 1) {
            for(string s: dp[i][arr[i]]) {
                answer.push_back(s);
            }
        }
    }
    
    sort(answer.begin(), answer.end());
    
    return answer;
}

File: 2021 카카오/메뉴 리뉴얼.java
import java.io.*;
import java.util.*;

class Solution {
    
    List<String> answerList = new ArrayList<>();
    Map<String, Integer> hashMap = new HashMap<>();
    
     public void combination(String order, String others, int cnt) {
        if(cnt == 0) {
            hashMap.put(order, hashMap.getOrDefault(order, 0) + 1);
            return;
        }
        
        for(int i = 0; i < others.length(); i++) {
            combination(order + others.charAt(i), others.substring(i + 1), cnt - 1);
        }
    }
    
    public String[] solution(String[] orders, int[] course) {       
        for(int i = 0; i < orders.length; i++) {
            char[] arr = orders[i].toCharArray();
            Arrays.sort(arr);
            orders[i] = String.valueOf(arr);
        }
        
        for(int c : course) {
            for(String o : orders) {
                combination("", o, c);
            }
            
            if(!hashMap.isEmpty()) {
                List<Integer> cntList = new ArrayList<>(hashMap.values());
                int maxN = Collections.max(cntList);
                
                if(maxN > 1) {
                    for(String key : hashMap.keySet()) {
                        if(hashMap.get(key) == maxN) {
                            answerList.add(key);
                        }
                    }
                }
                
                hashMap.clear();
            }
        }
        
        Collections.sort(answerList);
        String[] answer = new String[answerList.size()];
        
        for(int i = 0; i < answer.length; i++) {
            answer[i] = answerList.get(i);
        }
        
        return answer;
    }
}